File: lib/media/song.dart
import 'package:flutter_media_plugin/media/media.dart';
import 'package:flutter_media_plugin/utility.dart';

class Song implements Media {
  final String _key;
  final String _title;
  final String _artists;
  final String _album;

  // ignore: non_constant_identifier_names
  final String _album_art_url;
  final String _url;

  static String _capitalizeEveryWord(String str) {
    bool isSpaceFound = true;
    for (int i = 0; i < str.length; i++) {
      if (isSpaceFound) {
        str = str.substring(0, i) + str[i].toUpperCase() + str.substring(i + 1);
        isSpaceFound = false;
      }

      if (str[i] == ' ') {
        isSpaceFound = true;
      }
    }

    return str;
  }

  String get key => _key;

  String get title => _capitalizeEveryWord(_title);

  String get artists => _capitalizeEveryWord(_artists);

  String get album => _capitalizeEveryWord(_album);

  // ignore: non_constant_identifier_names
  String get album_art_url => _album_art_url;

  String get url => _url;

  Song(
      String key,
      String title,
      String artists,
      String album,
      // ignore: non_constant_identifier_names
      String album_art_url,
      String url)
      : _key = key,
        _title = title,
        _artists = artists,
        _album = album,
        _album_art_url = album_art_url,
        _url = url;

  @override
  Map<String, dynamic> toJson() {
    return {
      FieldUtility.song_key_tag: key,
      FieldUtility.song_title_tag: title,
      FieldUtility.song_artists_tag: artists,
      FieldUtility.song_album_tag: album,
      FieldUtility.song_album_art_url_tag: album_art_url,
      FieldUtility.song_url_tag: url,
    };
  }

  @override
  factory Song.fromMap(Map<String, dynamic> map) {
    String key = map[FieldUtility.song_key_tag];
    String title = map[FieldUtility.song_title_tag];
    String artists = map[FieldUtility.song_artists_tag];
    String album = map[FieldUtility.song_album_tag];
    String albumArtUrl = map[FieldUtility.song_album_art_url_tag];
    String url = map[FieldUtility.song_url_tag];
    Song song = Song(key, title, artists, album, albumArtUrl, url);
    return song;
  }
}

File: android/src/main/java/com/example/fluttermediaplugin/Media/Song.java
package com.example.fluttermediaplugin.Media;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;

import static com.example.fluttermediaplugin.Utility.MediaIds.ALBUM_ART_URL_TAG;
import static com.example.fluttermediaplugin.Utility.MediaIds.KEY_TAG;
import static com.example.fluttermediaplugin.Utility.MediaIds.SONG_ALBUM_TAG;
import static com.example.fluttermediaplugin.Utility.MediaIds.SONG_ARTISTS_TAG;
import static com.example.fluttermediaplugin.Utility.MediaIds.TITLE_TAG;
import static com.example.fluttermediaplugin.Utility.MediaIds.URL_TAG;

public final class Song implements Media {
    private String key;
    private String title;
    private String url;
    private String artists;
    private String album;
    private String album_art_url;

    private static String capitalizeEveryWord(String str) {
        boolean isSpaceFound = true;
        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);

            if(isSpaceFound) {
                String s = "" + c;
                s = s.toUpperCase();

                str = str.substring(0, i) + s + str.substring(i + 1);
                isSpaceFound = false;
            }
            if(c == ' ') {
                isSpaceFound = true;
            }
        }

        return str;
    }

    private Song(@NonNull String key, @NonNull String title, @NonNull String artist, @NonNull String album, @NonNull String album_art_url, @NonNull String url) {
        this.key = key;
        this.title = title;
        this.artists = artist;
        this.album = album;
        this.album_art_url = album_art_url;
        this.url = url;
    }

    @Override
    public String getKey() {
        return key;
    }

    @Override
    public String getTitle() {
        return capitalizeEveryWord(title);
    }

    @Override
    public String getUrl() {
        return url;
    }

    public String getAlbumArtUri() {
        return album_art_url;
    }

    public String getAlbum() {
        return capitalizeEveryWord(album);
    }

    public String getArtist() {
        return capitalizeEveryWord(artists);
    }

    @Override
    public Map<String, Object> toMap() {
        Map<String, Object> songMap = new HashMap<>();
        songMap.put(KEY_TAG, key);
        songMap.put(TITLE_TAG, title);
        songMap.put(SONG_ARTISTS_TAG, artists);
        songMap.put(SONG_ALBUM_TAG, album);
        songMap.put(ALBUM_ART_URL_TAG, album_art_url);
        songMap.put(URL_TAG, url);
        return songMap;
    }

    public static Song fromMap(Map<String, String> mapObject) {
        String key = mapObject.get(KEY_TAG);
        String title = mapObject.get(TITLE_TAG);
        String artists = mapObject.get(SONG_ARTISTS_TAG);
        String album = mapObject.get(SONG_ALBUM_TAG);
        String album_art_url = mapObject.get(ALBUM_ART_URL_TAG);
        String url = mapObject.get(URL_TAG);

        if(key == null || title == null || artists == null || album == null || album_art_url == null || url == null) {
            return null;
        }

        return new Song(key, title, artists, album, album_art_url, url);
    }
}
